[ConstraintSystem] Fix the ordering of functions with optional parame…

rudkx opened the pull request:
…ters.

Treat non-optional generic parameters as being more specialized than
optional generic parameters, and penalize any solutions that involve
generic arguments that are themselves Optional.

By doing these things, we can remove the special-cased code for the
two overloads of `??` in the stdlib, instead treating the `(T?, T)`
overload as better than the `(T?, T?)` overload except where a user
actually passes an optionally-typed value as the second parameter.

Fixes: rdar://problem/19748710

Commit:
[ConstraintSystem] Fix the ordering of functions with optional parameters.

Treat non-optional generic parameters as being more specialized than
optional generic parameters, and penalize any solutions that involve
generic arguments that are themselves Optional.

By doing these things, we can remove the special-cased code for the
two overloads of '??' in the stdlib, instead treating the (T?, T)
overload as better than the (T?, T?) overload except where a user
actually passes an optionally-typed value as the second parameter.

Fixes: rdar://problem/19748710

rudkx commented:
@swift-ci Please smoke test

jrose-apple commented on test/Constraints/diag_ambiguities.swift:
```diff
@@ -29,11 +29,12 @@ C(g) // expected-error{{ambiguous use of 'g'}}
-  return rdar29691909_callee(o) // expected-error{{ambiguous use of 'rdar29691909_callee'}}
+func rdar29691909(o: AnyObject) -> Int {
+  let r = rdar29691909_callee(o)
+  return r
```
This is a different test. Please leave the original one intact, where there's a same-type constraint.

Commit:
[ConstraintSystem] Ensure we prefer `f(_: Any)` to `f(_: Any?)` when both work.

...unless the argument is an `Any?`, in which case we prefer `f(_: Any?)`.

This change also results in our selecting f<T>(_: T) over f(_:
Any). Coercing with 'as Any' makes it possible to explicitly select
the Any overload. Previously there was no way to select the generic
overload.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
As my latest test update shows, the change for `Any` vs. `Any?` introduces another change, which is that we now prefer `f<T>(x: T)` over `f(Any)`, which seems to me like a reasonable choice since you can select the second overload by doing `f(arg as Any)` but under the old scheme you can never select the `f<T>(x: T)` variant.

@DougGregor what do you think?

rudkx commented:
@swift-ci Please test source compatibility
